tools/draincluster/drain/drain.go

```diff
@@ -0,0 +1,244 @@
+				validCondition.Reason == condition.EvictionInvalidMissingCRBMessage {
+				continue
+			}
+		}
+		executedCondition := eviction.GetCondition(string(placementv1beta1.PlacementEvictionConditionTypeExecuted))
+		if executedCondition == nil || executedCondition.Status == metav1.ConditionFalse {
+			isDrainSuccessful = false
+			log.Printf("eviction %s was not executed successfully for CRP %s", evictionName, crpName)
+			continue
+		}
+		log.Printf("eviction %s was executed successfully for CRP %s", evictionName, crpName)
+		// log each resource evicted by CRP.
+		for i := range h.ClusterResourcePlacementResourcesMap[crpName] {
+			resourceIdentifier := h.ClusterResourcePlacementResourcesMap[crpName][i]
+			log.Printf("evicted resource %s propagated by CRP %s", fmt.Sprintf(resourceIdentifierKeyFormat, resourceIdentifier.Group, resourceIdentifier.Version, resourceIdentifier.Kind, resourceIdentifier.Name, resourceIdentifier.Namespace), crpName)
+		}
+	}
+
+	return isDrainSuccessful, nil
+}
+
+func (h *Helper) cordon(ctx context.Context) error {
+	// add taint to member cluster to ensure resources aren't scheduled on it.
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		var mc clusterv1beta1.MemberCluster
+		if err := h.HubClient.Get(ctx, types.NamespacedName{Name: h.ClusterName}, &mc); err != nil {
+			return err
+		}
+
+		// search to see cordonTaint already exists on the cluster.
+		for i := range mc.Spec.Taints {
+			if mc.Spec.Taints[i] == toolsutils.CordonTaint {
+				return nil
+			}
+		}
+
+		// add taint to member cluster to cordon.
+		mc.Spec.Taints = append(mc.Spec.Taints, toolsutils.CordonTaint)
+
+		return h.HubClient.Update(ctx, &mc)
+	})
+}
+
+func (h *Helper) fetchClusterResourcePlacementToEvict(ctx context.Context) error {
+	var crbList placementv1beta1.ClusterResourceBindingList
+	if err := h.HubClient.List(ctx, &crbList); err != nil {
+		return fmt.Errorf("failed to list cluster resource bindings: %w", err)
+	}
+
+	// find all unique CRP names for which eviction needs to occur.
+	for i := range crbList.Items {
+		crb := crbList.Items[i]
+		var targetBinding *placementv1beta1.ClusterResourceBinding
+		if crb.Spec.TargetCluster == h.ClusterName {
+			targetBinding = &crb
+		}
+
+		if targetBinding != nil {
+			err := wait.ExponentialBackoffWithContext(ctx, retry.DefaultBackoff, func(ctx context.Context) (bool, error) {
+				if getErr := h.HubClient.Get(ctx, types.NamespacedName{Name: targetBinding.Name}, targetBinding); getErr != nil {
+					// binding may have been deleted.
+					if k8errors.IsNotFound(getErr) {
+						return true, nil
+					}
+					return false, getErr
+				}
+				if !evictionutils.IsPlacementPresent(targetBinding) {
+					// need to wait until placement is present on member cluster.
+					return false, nil
+				}
+				return true, nil
+			})
+
+			if err != nil {
+				return fmt.Errorf("failed to wait for placement to be present on member cluster: %w", err)
```

Comment:
Hi Arvind! This is a bit weird too if I read it correctly: if a binding no the cluster cannot get ready (i.e., rollout is stuck), the cluster cannot be drained?